File: array/PrefixSum2D/python/PrefixSum2D.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Sequence


@dataclass(frozen=True)
class RectQuery:
    r1: int
    c1: int
    r2: int
    c2: int


class PrefixSum2D:
    def __init__(self, matrix: Sequence[Sequence[int]]) -> None:
        rows = len(matrix)
        cols = len(matrix[0]) if rows else 0
        self._prefix = [[0] * (cols + 1) for _ in range(rows + 1)]

        for r in range(rows):
            for c in range(cols):
                self._prefix[r + 1][c + 1] = (
                    matrix[r][c]
                    + self._prefix[r][c + 1]
                    + self._prefix[r + 1][c]
                    - self._prefix[r][c]
                )

    def rectangle_sum(self, r1: int, c1: int, r2: int, c2: int) -> int:
        return (
            self._prefix[r2 + 1][c2 + 1]
            - self._prefix[r1][c2 + 1]
            - self._prefix[r2 + 1][c1]
            + self._prefix[r1][c1]
        )


def answer_rectangle_queries(
    matrix: Sequence[Sequence[int]], queries: Iterable[RectQuery]
) -> list[int]:
    prefix_sum = PrefixSum2D(matrix)
    return [
        prefix_sum.rectangle_sum(query.r1, query.c1, query.r2, query.c2)
        for query in queries
    ]

File: array/PrefixSum2D/rust/src/prefix_sum_2d.rs
#[derive(Clone, Copy)]
pub struct RectQuery {
    pub r1: usize,
    pub c1: usize,
    pub r2: usize,
    pub c2: usize,
}

pub struct PrefixSum2D {
    prefix: Vec<Vec<i64>>,
}

impl PrefixSum2D {
    pub fn new(matrix: &[Vec<i64>]) -> Self {
        let rows = matrix.len();
        let cols = if rows == 0 { 0 } else { matrix[0].len() };
        let mut prefix = vec![vec![0_i64; cols + 1]; rows + 1];

        for r in 0..rows {
            for c in 0..cols {
                prefix[r + 1][c + 1] = matrix[r][c]
                    + prefix[r][c + 1]
                    + prefix[r + 1][c]
                    - prefix[r][c];
            }
        }

        Self { prefix }
    }

    pub fn rectangle_sum(&self, r1: usize, c1: usize, r2: usize, c2: usize) -> i64 {
        self.prefix[r2 + 1][c2 + 1]
            - self.prefix[r1][c2 + 1]
            - self.prefix[r2 + 1][c1]
            + self.prefix[r1][c1]
    }
}

pub fn answer_rectangle_queries(matrix: &[Vec<i64>], queries: &[RectQuery]) -> Vec<i64> {
    let prefix_sum = PrefixSum2D::new(matrix);
    queries
        .iter()
        .map(|query| prefix_sum.rectangle_sum(query.r1, query.c1, query.r2, query.c2))
        .collect()
}
